cmake/grpc.cmake

```diff
@@ -11,39 +11,115 @@ unset(_grpc_cxx_standard)
 unset(_package_providers)
 unset(_patch_clause)
 
-function(prepare_grpc_patch_step OUTVAR)
-  if(NOT DEFINED GRPC_VERSION OR GRPC_VERSION STREQUAL "")
-    message(WARNING "GRPC_VERSION undefined; cannot select patch file")
+#-----------------------------------------------------------------------
+# define_grpc_patch_file()
+# Returns the filename of the template patch file.
+#
+# Parameters:
+#   PATCHDIR - directory containing patch files.
+#   OUTVAR - name of output variable.
+#
+# Variables:
+#   GRPC_PATCH - name of gRPC patch file.
+#   GRPC_VERSION - version of gRPC being patched.
+#
+# Choices are, in descending order of precedence:
+#  1. ${GRPC_PATCHFILE}
+#  2. grpc-v$(GRPC_VERSION}.patch.in
+#  3. grpc-current.patch.in
+#
+# Returns an empty string if we cannot determine a valid patch file name.
+#-----------------------------------------------------------------------
+function(define_grpc_patch_file PATCHDIR OUTVAR)
+  # Specified patch file
+  if(DEFINED GRPC_PATCHFILE AND NOT GRPC_PATCHFILE STREQUAL "")
+    set(patchfile "${GRPC_PATCHFILE}")
+    if(EXISTS ${PATCHDIR}/${patchfile})
+      set(${OUTVAR} "${patchfile}" PARENT_SCOPE)
+    else()
+      message("----------------------------------------")
+      message("Patch file '${patchfile}' does not exist.")
+      message("Check definition of GRPC_PATCHFILE.")
+      message("Its current value is '${GRPC_PATCHFILE}.'")
+      message("----------------------------------------")
+      message(WARNING "Skipping patch stage")
+      set(${OUTVAR} "" PARENT_SCOPE)
+    endif()
     return()
   endif()
 
-  set(patchfile grpc-v${GRPC_VERSION}.patch.in)
-  set(patchdir ${CMAKE_SOURCE_DIR}/cmake/patches)
+  # Patch file for gRPC version
+  if(DEFINED GRPC_VERSION)
+    set(version_file "grpc-v${GRPC_VERSION}.patch.in")
+    if(EXISTS ${PATCHDIR}/${version_file})
+      set(${OUTVAR} "${version_file}" PARENT_SCOPE)
+      return()
+    endif()
+  endif()
 
-  if(NOT EXISTS ${patchdir}/${patchfile})
-    message(WARNING "Patch file for gRPC version ${GRPC_VERSION} not found")
+  # Current patch file
+  set(patchfile "grpc-current.patch.in")
```

Comment:
Is there a risk of having a patch file called `grpc-current` and being leftover in a user's build system after a grpc upgrade, resulting in patching mismatch between grpc version and patch file that should be used?